[MP Rewrite Part 1] Break out and clean up BallState

kylestach opened the pull request:
This substantially cleans up the ball state code, and moves it from
`SystemState` (which we are moving away from in general) to `WorldState`
(which is restricted to physics-only information).

kylestach committed:
[MP Rewrite Part 1] Break out and clean up BallState

This substantially cleans up the ball state code, and moves it from
`SystemState` (which we are moving away from in general) to `WorldState`
(which is restricted to physics-only information.

kylestach committed:
Merge branch 'mp-rewrite/base' into mp-rewrite/ball-state

kylestach committed:
Missed a few files

oswinso commented on soccer/WorldState.hpp:
```diff
@@ -14,12 +20,80 @@ struct RobotState {
 };
 
+/**
+ * Our belief about the ball's current position and velocity.
```
`\brief`

oswinso commented on soccer/WorldState.hpp:
```diff
@@ -14,12 +20,80 @@ struct RobotState {
+    BallState() = default;
+
+    /**
+     * Construct a BallState with a valid estimate.
```
`\brief` for all of these?

oswinso commented on soccer/WorldState.hpp:
```diff
@@ -14,12 +20,80 @@ struct RobotState {
+    [[nodiscard]] BallState predict_at(RJ::Time time) const;
+
+    /**
+     * Similar to \ref predict_at "predict_at(RJ::Time)", but for a duration in
+     * the future
```
Clarify by saying it's `timestamp + seconds`

oswinso commented on soccer/WorldState.hpp:
```diff
@@ -1,7 +1,13 @@
+#include "planning/Trajectory.hpp"
+#include "time.hpp"
+
+// TODO(Kyle): Make this configurable
```
Issue number pls

kylestach replied:
#1498

oswinso commented on soccer/gameplay/robocup-py.cpp:
```diff
@@ -904,21 +1007,20 @@ BOOST_PYTHON_MODULE(robocup) {
            bases<Robot>>("OpponentRobot", init<Context*, int>());
 
-    class_<Ball, std::shared_ptr<Ball>>("Ball", init<>())
+    class_<BallState, BallState*>("Ball", init<>())
```
Why the change from `shared_ptr` to raw ptr?

kylestach replied:
We no longer have a reason to use `BallState` with `shared_ptr` values anywhere (it's a tiny little value type, there's no reason to avoid copies).

oswinso reviewed (changes requested):
A few small things

JNeiger commented on soccer/WorldState.hpp:
```diff
@@ -14,12 +20,80 @@ struct RobotState {
+        Geometry2d::Point* out = nullptr) const;
+
+    /**
+     * Query the time before the ball goes a certain distance.
```
Not: param and return descriptions. Specially the optional return

JNeiger commented on soccer/gameplay/robocup-py.cpp:
```diff
@@ -144,19 +170,18 @@ void OurRobot_move_to_end_vel(OurRobot* self, Geometry2d::Point* endPos,
 
-void OurRobot_settle(OurRobot* self) { self->settle(std::nullopt); }
-
-void OurRobot_settle_w_bounce(OurRobot* self, Geometry2d::Point* bounceTarget) {
```
This removes the one touch functionality in the path planned layer. We'd have to revert back to the previous version, which I think just used move

kylestach replied:
Reverted.

JNeiger commented on soccer/planning/DynamicObstacle.hpp:
```diff
@@ -1,54 +1,13 @@
-#include "planning/MotionInstant.hpp"
-#include "Utils.hpp"
-#include "Geometry2d/Shape.hpp"
+#include <memory>
```
Is this include needed?

JNeiger commented on soccer/planning/DynamicObstacle.hpp:
```diff
@@ -1,54 +1,13 @@
-    }
+class Trajectory;
+struct DynamicObstacle {
+    DynamicObstacle(const Geometry2d::Circle& circle, const Trajectory* path)
```
Need some comments on what all this represents and how to use this

JNeiger replied:
Circles have a center and radius, are there any specifics on the center?

kylestach replied:
Clarified by making this only take a radius instead of a real circle object.

JNeiger commented on soccer/BallTest.cpp:
```diff
@@ -0,0 +1,45 @@
+    Geometry2d::Point actual;
+    RJ::Seconds t = state.query_seconds_near(Geometry2d::Point(5, 0), &actual);
+
+    // TODO: For some reason query_stop_time and query_seconds_to_dist give
```
Issue #?

kylestach committed:
Address review comments, improve BallState docs

kylestach committed:
Update comments, re-add settle_w_bounce

JNeiger reviewed (commented):
Minus the circle comment, this is good to merge

kylestach committed:
Make DynamicObstacle take only a radius to enforce trajectory-centering.

oswinso reviewed (approved):
lgtm

JNeiger reviewed (approved):
Lgtm

egordon9dev commented on soccer/gameplay/plays/defense/DefendClear.py:
```diff
@@ -0,0 +1,35 @@
+import standard_play
```
These plays files might have been added on accident. Is this supposed to be here?

egordon9dev commented on soccer/WorldState.cpp:
```diff
@@ -0,0 +1,131 @@
+
+    // The trajectory interface fits cubic splines. Luckily, a cubic spline
+    // between two instants that can be connected by a constant acceleration
+    // (like we have here) will be, and so we can use this for our trajectory.
```
This is pretty clever. nice.

kylestach committed:
Remove accidentally added defense files

kylestach committed:
[MP Rewrite Part 2] Introduce the Trajectory abstraction (#1494)

A Trajectory is a piecewise-cubic function interpolating between a
vector of time-indexed RobotInstant waypoints.

This is not expected to pass CI.

* Add motion control changes for following Planning::Trajectory

* Address review comments

* Change fixAngleRadians to be more efficient

* Address review comments

* Tidy up != operators (use == in defn.)

* [MP Rewrite Part 3] Add PlanRequest and the planner node. (#1495)

* Add PlanRequest and the planner node.

This also overhauls the MotionCommand structure.

* Remove dead code from Processor, replace with TODO

* Make path target goal a linear motion instant

* Address review comments

[MP Rewrite Part 4] Motion Planning Primitives (#1496)

Primitives for motion planning including angle planning, RRT, velocity
profiling, Bezier fitting, and replanning. This is anticipated to be the
largest portion of the motion planning rewrite.

This is not expected to pass CI.

* Remove disabled for-loop in Processor

* Fix TODOs and docs in planning primitives

* Address review comments

* Make off-path threshold configurable

[MP Rewrite Part 5] Port previous planners to new framework (#1497)

Most of these are not functional changes from the old planners, but
rather simple ports to the new primitives. The biggest changes are that
we now use `Replanner` static functions and manually manage our own
"previous trajectory" objects instead of having "behavior-level"
planners (i.e. settle) own instances of an RRT planner and interacting
with it through plan requests.

This is the final PR in the rewrite, and it should pass CI. At the point
where it passes CI and all PRs are ready for merge, each PR should be
merged into its parent in sequence.

* Address review comments

* Move applyLowPassFilter into common, add TODOs

* Enable escape-obstacles planner

[MP Rewrite Part 6] Cleanup and fixes (#1512)

* Cleanup and fixes to make code run properly

* Fix test compilation and disable a trapezoidal motion profile edge case.

Some planner tests are still failing. This will be addressed in a later
commit.

* Prettify

* Make e2e planner tests pass.

Because they generate random scenarios, and the RRT planner is not
guaranteed to converge in a finite number of iterations (and the Bezier
approximation may be incorrect), these tests are currently flaky. That
should be fixed before this is merged to staging.

* Tidy fixes

* Linting

* Formatting

* Remove deleted file from CMakeLists, remove reuse condition from settle

* Don't replan in Collect::control

* Optimizations to geometry methods

* Reset and regenerate sim config

This deletes unnecessary variables and moves planning-related variables
into PathPlanner

* Formatting

* Add TODO in EscapeObstacles

* Tidy fixes

* Revert inconsistent clang-tidy recommended fix

* Temporarily disable flaky planner test

* Add NOLINTNEXTLINE

* Remove old sim config and add TODOs for trajectory utils
